src/scope.ml

```diff
@@ -195,7 +284,11 @@ let scope : mode -> sig_state -> env -> p_term -> tbox = fun md ss env t ->
         in
         _TEnv (Bindlib.box_var x) (Array.map (scope env) ts)
     | (P_Patt(_,_)     , _        ) -> fatal t.pos "Only allowed in rules."
-    | (P_Appl(t,u)     , _        ) -> _Appl (scope env t) (scope env u)
+    | (P_Appl(t,u)     , _        ) -> 
+          let (f, args) = splitFunArgs (none (P_Appl(t,u))) in
```

Comment:
Replace "none (P_Appl(t,u))" by "t".

Reply:
splitFunArgs is called at each P_Appl node. This doesn't seem very efficient. splitFunArgs should be called on t.elt before entering the match on (t.elt, md).

Reply:
splitFunArgs returns the symbol being applied (the function) and all its arguments. If I call splitFunArgs just on [t], instead of [P_Appl(t,u)], I won't get the last argument of the call, namely [u], and will need to add it afterwards.
Anyway, this portion of the code is probably going to move.

Reply:
I was talking about the outer t.